sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/BufferingStreamObserver.java

```diff
@@ -0,0 +1,166 @@
+ * are expected to advance the {@link Phaser} whenever the underlying {@link CallStreamObserver}
+ * becomes ready.
+ */
+@ThreadSafe
+public final class BufferingStreamObserver<T> implements StreamObserver<T> {
+  private static final Object POISON_PILL = new Object();
+  private final LinkedBlockingDeque<T> queue;
+  private final Phaser phaser;
+  private final CallStreamObserver<T> outboundObserver;
+  private final Future<?> queueDrainer;
+  private final int bufferSize;
+
+  public BufferingStreamObserver(
+      Phaser phaser,
+      CallStreamObserver<T> outboundObserver,
+      ExecutorService executor,
+      int bufferSize) {
+    this.phaser = phaser;
+    this.bufferSize = bufferSize;
+    this.queue = new LinkedBlockingDeque<>(bufferSize);
+    this.outboundObserver = outboundObserver;
+    this.queueDrainer = executor.submit(this::drainQueue);
+  }
+
+  private void drainQueue() {
+    try {
+      while (true) {
+        int currentPhase = phaser.getPhase();
+        while (outboundObserver.isReady()) {
+          T value = queue.take();
+          if (value != POISON_PILL) {
+            outboundObserver.onNext(value);
+          } else {
+            return;
+          }
+        }
+        phaser.awaitAdvance(currentPhase);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Override
+  public void onNext(T value) {
+    try {
+      // Attempt to add an element to the bounded queue occassionally checking to see
+      // if the queue drainer is still alive.
+      while (!queue.offer(value, 60, TimeUnit.SECONDS)) {
+        if (queueDrainer.isDone()) {
+          throw new IllegalStateException("Stream observer has finished.");
+        }
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    synchronized (outboundObserver) {
+      // If we are done, then a previous caller has already shutdown the queue processing thread
+      // hence we don't need to do it again.
+      if (!queueDrainer.isDone()) {
+        // We check to see if we were able to successfully insert the poison pill at the front of
+        // the queue to cancel the processing thread eagerly or if the processing thread is done.
+        try {
+          // The order of these checks is important because short circuiting will cause us to
+          // insert into the queue first and only if it fails do we check that the thread is done.
+          while (!queue.offerFirst((T) POISON_PILL, 60, TimeUnit.SECONDS)
+              || !queueDrainer.isDone()) {
+          }
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(e);
+        }
+        waitTillFinish();
+      }
```

Comment:
I also agree but I tried twice before during implementation and everything I came up with I felt was strictly worse. Vikas also mentioned the same thing so if you want to take a stab at it, the improvement would be good.